File: src/main/java/at/searles/parsing/combinators/ParserThenRecognizer.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;

public class ParserThenRecognizer<T> implements Parser<T>, Recognizer.Then {
    private final Parser<T> left;
    private final Recognizer right;
    private final boolean allowParserBacktrack;

    public ParserThenRecognizer(Parser<T> left, Recognizer right, boolean allowParserBacktrack) {
        this.left = left;
        this.right = right;
        this.allowParserBacktrack = allowParserBacktrack;
    }

    @Override
    public Recognizable left() {
        return left;
    }

    @Override
    public Recognizable right() {
        return right;
    }

    @Override
    public boolean allowParserBacktrack() {
        return allowParserBacktrack;
    }

    @Override
    public T parse(ParserStream stream) {
        long offset = stream.getOffset();

        // to restore if backtracking
        long preStart = stream.getStart();
        long preEnd = stream.getEnd();

        T result = left.parse(stream);

        if (result == null) {
            return null;
        }

        // The start position of left.
        long start = stream.getStart();

        if (!right.recognize(stream)) {
            if(stream.getOffset() != offset) {
                throwIfNoBacktrack(stream);
                stream.backtrackToOffset(offset);
            }

            stream.setStart(preStart);
            stream.setEnd(preEnd);
            return null;
        }

        stream.setStart(start);

        return result;
    }

    @Override
    public ConcreteSyntaxTree print(T t) {
        ConcreteSyntaxTree output = left.print(t);

        // Recognizer.printTo always succeeds.
        return output != null ? output.consRight(right.print()) : null;
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/ParserThenRecognizer.kt
package at.searles.parsing.combinators

import at.searles.parsing.Parser
import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizable.Then
import at.searles.parsing.Recognizer
import at.searles.parsing.printing.ConcreteSyntaxTree

class ParserThenRecognizer<T>(override val left: Parser<T>, override val right: Recognizer) : Parser<T>, Then {
    override fun parse(stream: ParserStream): T? {
        val offset = stream.offset

        // to restore if backtracking
        val preStart = stream.start
        val preEnd = stream.end
        val result = left.parse(stream) ?: return null

        // The start position of left.
        val start = stream.start

        if (!right.recognize(stream)) {
            if (stream.offset != offset) {
                stream.requestBacktrackToOffset(this, offset)
            }

            stream.start = preStart
            stream.end = preEnd
            return null
        }

        stream.start = start
        return result
    }

    override fun print(item: T): ConcreteSyntaxTree? {
        val output = left.print(item)

        // Recognizer.printTo always succeeds.
        return output?.consRight(right.print())
    }

    override fun toString(): String {
        return createString()
    }
}